Improve the trim incompatibility message for Assembly.GetType

vitek-karas opened the pull request:
Currently calling `Assembly.GetType` is always a warning when trimming. But typically the code can be changed to use `Type.GetType` with assembly qualified type name, which is a trim compatible pattern.

This modifies the warning message on the `RequiresUnreferencedCode` attribute for `Assembly.GetType` to suggest using `Type.GetType` instead.

vitek-karas committed:
Improve the trim incompatibility message for Assembly.GetType

Currently calling `Assembly.GetType` is always a warning when trimming. But typically the code can be changed to use `Type.GetType` with assembly qualified type name, which is a trim compatible pattern.

This modifies the warning message on the `RequiresUnreferencedCode` attribute for `Assembly.GetType` to suggest using `Type.GetType` instead.

eerhardt commented:
Don't you need to update the `ref` assembly code as well?

vitek-karas committed:
Fix ref assembly

vitek-karas commented:
> Don't you need to update the ref assembly code as well?

Of course - I was... dumb. Should be fixed now.

MichalStrehovsky commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:
```diff
@@ -242,7 +242,7 @@ public override AssemblyName GetName(bool copiedName)
         }
 
-        [RequiresUnreferencedCode("Types might be removed")]
+        [RequiresUnreferencedCode("Types might be removed by trimming. Consider using Type.GetType instead.")]
```
If we're updating CoreCLR's RuntimeAssembly, should we also update Mono's and NativeAOT's?

vitek-karas replied:
Done - thanks for reminding me

MichalStrehovsky reviewed (commented):
If we want to be prescriptive in the message, should we change this to `Types might be removed by trimming. If the type name is a string literal, consider using Type.GetType instead.`.

Basically the advice to use `Type.GetType` is not useful if this is not a string constant that we can analyze or is properly annotated.

vitek-karas committed:
PR feedback

Improve the message.

Use the new message in mono classes as well (to unify the message across all implementations of the abstract methods)

vitek-karas commented:
Thanks for the suggestion @MichalStrehovsky  - I changed the message everywhere.

MichalStrehovsky reviewed (approved):
Thank you!
